openhcl/underhill_mem/src/mapping.rs

```diff
@@ -17,22 +17,208 @@ use memory_range::MemoryRange;
 use parking_lot::Mutex;
 use sparse_mmap::SparseMapping;
 use std::ptr::NonNull;
+use std::sync::Arc;
 use thiserror::Error;
 use vm_topology::memory::MemoryLayout;
 
+pub struct GuestPartitionMemoryBuilder<'a> {
+    memory_layout: &'a MemoryLayout,
+    valid_memory: Option<Arc<GuestValidMemory>>,
+}
+
+impl<'a> GuestPartitionMemoryBuilder<'a> {
+    /// If `valid_bitmap_state` is `Some`, a bitmap is created to track the
+    /// accessibility state of each page in the lower VTL memory. The bitmap is
+    /// initialized to the provided state.
+    ///
+    /// This is used to support tracking the shared/encrypted state of each
+    /// page.
+    pub fn new(
+        memory_layout: &'a MemoryLayout,
+        valid_bitmap_state: Option<bool>,
+    ) -> Result<Self, MappingError> {
+        let valid_memory = valid_bitmap_state
+            .map(|state| GuestValidMemory::new(memory_layout, state).map(Arc::new))
+            .transpose()?;
+        Ok(Self {
+            memory_layout,
+            valid_memory,
+        })
+    }
+
+    /// Returns the built partition-wide valid memory.
+    pub fn partition_valid_memory(&self) -> Option<Arc<GuestValidMemory>> {
+        self.valid_memory.clone()
+    }
+
+    /// Build a [`GuestMemoryMapping`], feeding in any related partition-wide
+    /// state.
+    pub fn build_guest_memory_mapping(
+        &self,
+        mshv_vtl_low: &MshvVtlLow,
+        memory_mapping_builder: &mut GuestMemoryMappingBuilder,
+    ) -> Result<GuestMemoryMapping, MappingError> {
+        memory_mapping_builder
+            .use_partition_valid_memory(self.valid_memory.clone())
+            .build(mshv_vtl_low, self.memory_layout)
```

Comment:
nit: I think I'd prefer that _this_ method be private and update `GuestMemoryMappingBuilder::build` to take a `&GuestPartitionMemoryBuilder<'_>`.

This thing isn't really a builder, after all. It's just memory layout and an optional `GuestValidMemory` bound together.